go-controller/pkg/ovn/egressgw.go

```diff
@@ -585,48 +571,272 @@ func (oc *Controller) extSwitchPrefix(nodeName string) (string, error) {
+			for _, ovnRoute := range ovnRoutes {
+				gr := strings.TrimPrefix(ovnRoute.router, types.GWRouterPrefix)
+				if err := oc.delHybridRoutePolicyForPod(net.ParseIP(podIP), gr, true); err != nil {
+					klog.Errorf("Error while removing hybrid policy for pod IP: %s, on node: %s, error: %v",
+						podIP, gr, err)
+				}
+			}
+		}
+	}
+}
+
+func getExGwPodIPs(gatewayPod *kapi.Pod) ([]net.IP, error) {
+	var foundGws []net.IP
+	if gatewayPod.Annotations[routingNetworkAnnotation] != "" {
+		var multusNetworks []nettypes.NetworkStatus
+		err := json.Unmarshal([]byte(gatewayPod.ObjectMeta.Annotations[nettypes.NetworkStatusAnnot]), &multusNetworks)
+		if err != nil {
+			return nil, fmt.Errorf("unable to unmarshall annotation k8s.v1.cni.cncf.io/network-status on pod %s: %v",
+				gatewayPod.Name, err)
+		}
+		for _, multusNetwork := range multusNetworks {
+			if multusNetwork.Name == gatewayPod.Annotations[routingNetworkAnnotation] {
+				for _, gwIP := range multusNetwork.IPs {
+					ip := net.ParseIP(gwIP)
+					if ip != nil {
+						foundGws = append(foundGws, ip)
+					}
+				}
+			}
+		}
+	} else if gatewayPod.Spec.HostNetwork {
+		for _, podIP := range gatewayPod.Status.PodIPs {
+			ip := net.ParseIP(podIP.IP)
+			if ip != nil {
+				foundGws = append(foundGws, ip)
+			}
+		}
+	} else {
+		klog.Errorf("Ignoring pod %s as an external gateway candidate. Invalid combination "+
+			"of host network: %t and routing-network annotation: %s", gatewayPod.Name, gatewayPod.Spec.HostNetwork,
+			gatewayPod.Annotations[routingNetworkAnnotation])
+		return nil, nil
+	}
+	return foundGws, nil
+}
+
+func (oc *Controller) buildClusterECMPCacheFromNamespaces(clusterRouteCache map[string][]string) {
+	namespaces, err := oc.watchFactory.GetNamespaces()
 	if err != nil {
-		klog.Errorf("cleanECMPRoutes: failed to list ecmp routes %v %s", err, stderr)
+		klog.Errorf("Error getting all namespaces for exgw ecmp route sync: %v", err)
 		return
 	}
+	for _, namespace := range namespaces {
+		if _, ok := namespace.Annotations[routingExternalGWsAnnotation]; !ok {
+			continue
+		}
+		// namespace has exgw routes, build cache
+		gwIPs, err := parseRoutingExternalGWAnnotation(namespace.Annotations[routingExternalGWsAnnotation])
+		if err != nil {
+			klog.Errorf("Unable to clean ExGw ECMP routes for namespace: %s, %v", namespace.Name, err)
+			continue
+		}
+		// get all pods in the namespace
+		nsPods, err := oc.watchFactory.GetPods(namespace.Name)
+		if err != nil {
+			klog.Errorf("Unable to clean ExGw ECMP routes for namespace: %s, %v",
+				namespace, err)
+			continue
+		}
+		for _, gwIP := range gwIPs {
+			for _, nsPod := range nsPods {
+				for _, podIP := range nsPod.Status.PodIPs {
+					if utilnet.IsIPv6(gwIP) != utilnet.IsIPv6String(podIP.IP) {
+						continue
+					}
+					if _, ok := clusterRouteCache[podIP.String()]; ok {
+						clusterRouteCache[podIP.IP] = append(clusterRouteCache[podIP.IP], gwIP.String())
```

Comment:
```suggestion
						if gwIP.String() not in clusterRouteCache[podIP.IP]:
						    clusterRouteCache[podIP.IP] = append(clusterRouteCache[podIP.IP], gwIP.String())
```

So the problem here is first we do `buildClusterECMPCacheFromPods` which would already add the exgw serving pod IPs to the cache if the `routing-namespace` annotation is set. Then again here we would re-add the same gwIPs if the namespace has the `routing-external-gws` annotation is present. We'd keep adding the same IPs over,

Sample output on my KIND:

```
I0810 10:14:57.528872      37 egressgw.go:845] Cluster ECMP route cache is: map[10.244.1.6:[172.18.0.3 172.18.0.3 172.18.0.4] 10.244.2.5:[172.18.0.3 172.18.0.3 172.18.0.4]]
```

Reply:
Thanks Surya. Can you fix the suggestion. It looks like python

Reply:
yea let me get this right into the suggested change (will do it in a bit), then you can accept all the suggested changes and I think we are good to go.

Reply:
ok done. I've tested the new patch ^ on KIND:

```
I0811 06:56:36.898095      38 egressgw.go:714] Cluster ECMP route cache is: map[10.244.1.4:[172.18.0.4 172.18.0.2] 10.244.2.4:[172.18.0.4 172.18.0.2]]
```

Looks good.